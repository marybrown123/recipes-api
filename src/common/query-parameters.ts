import { ApiProperty } from '@nestjs/swagger';
import { Transform } from 'class-transformer';
import { IsNumber, IsOptional, IsString } from 'class-validator';

export class QueryParameters {
  @IsString()
  @IsOptional()
  @ApiProperty({ example: 'dumplings', type: 'string' })
  name?: string;

  @Transform(({ value }) => (value ? parseInt(value) : null))
  @IsNumber()
  @IsOptional()
  @ApiProperty({ example: 1, type: 'number' })
  page = 1;

  @Transform(({ value }) => (value ? parseInt(value) : null))
  @IsNumber()
  @IsOptional()
  @ApiProperty({ example: 3, type: 'number' })
  limit = 3;
}
