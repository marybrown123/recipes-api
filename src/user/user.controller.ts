import { Body, Controller, Post } from '@nestjs/common';
import * as bcrypt from 'bcrypt';
import { UserService } from './user.service';
import { CreateUserDTO } from './DTOs/create-user.DTO';
@Controller('/user')
export class UserController {
  constructor(private readonly usersService: UserService) {}

  @Post('/signup')
  async createUser(@Body() user: CreateUserDTO) {
    const hashedPassword = await bcrypt.hash(user.password, 10);
    const userForDb = {
      name: user.name,
      password: hashedPassword,
    };
    const createdUser = await this.usersService.createUser(userForDb);
    return createdUser;
  }
}
