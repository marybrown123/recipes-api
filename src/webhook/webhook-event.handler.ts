import { HttpService } from '@nestjs/axios';
import { Injectable } from '@nestjs/common';
import { Interval } from '@nestjs/schedule';
import { Recipe, User, WebhookEventStatus } from '@prisma/client';
import { lastValueFrom } from 'rxjs';
import { PrismaService } from '../prisma/prisma.service';
import { WebhookName } from './enums/webhookName.enum';
import { WebhookService } from './webhook.service';

@Injectable()
export class WebhookEventHandler {
  constructor(
    private readonly prismaService: PrismaService,
    private readonly httpService: HttpService,
    private readonly webhookService: WebhookService,
  ) {}
  async createWebhookEvent(
    webhookName: WebhookName,
    payload: Recipe | User,
  ): Promise<void> {
    const webhook = await this.webhookService.fetchOneWebhookByName(
      webhookName,
    );

    await this.prismaService.webhookEvent.create({
      data: {
        webhook: {
          connect: {
            id: webhook.id,
          },
        },
        payload,
        retriesLeft: webhook.retriesAmount,
      },
    });
  }

  @Interval(5000)
  async triggerWebhookEvents() {
    const webhookEvents = await this.prismaService.webhookEvent.findMany({
      where: { status: WebhookEventStatus.Pending },
      include: { webhook: true },
    });

    if (webhookEvents) {
      webhookEvents.forEach(async (webhookEvent) => {
        if (webhookEvent.webhook.isEnabled) {
          try {
            await this.prismaService.webhookEvent.update({
              where: { id: webhookEvent.id },
              data: { retriesLeft: webhookEvent.retriesLeft - 1 },
            });
            await lastValueFrom(
              this.httpService.post(
                webhookEvent.webhook.url,
                webhookEvent.payload,
              ),
            );
            await this.updateWebhookEventStatus(
              webhookEvent.id,
              WebhookEventStatus.Succed,
            );
          } catch (error) {
            const webhookEventFromDb =
              await this.prismaService.webhookEvent.findFirst({
                where: { id: webhookEvent.id },
              });
            if (webhookEventFromDb.retriesLeft === 0) {
              await this.updateWebhookEventStatus(
                webhookEvent.id,
                WebhookEventStatus.Failed,
              );
            }
          }
        }
      });
    }
  }

  async updateWebhookEventStatus(
    webhookEventId: number,
    status: WebhookEventStatus,
  ) {
    await this.prismaService.webhookEvent.update({
      where: { id: webhookEventId },
      data: { status },
    });
  }
}
