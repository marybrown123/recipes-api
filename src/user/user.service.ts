import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { PrismaService } from 'src/prisma/prisma.service';
import { CreateUserDTO } from './DTOs/create-user.DTO';
import { Role } from '@prisma/client';
import { UserResponse } from './responses/user.response';
import * as bcrypt from 'bcrypt';

@Injectable()
export class UserService {
  constructor(private prisma: PrismaService) {}

  async createUser(user: CreateUserDTO) {
    const userFromDb = await this.prisma.user.findUnique({
      where: { name: user.name },
    });

    if (userFromDb) {
      throw new HttpException('User already exists', HttpStatus.BAD_REQUEST);
    }

    const hashedPassword = await this.hashPassword(user.password);

    const userForDb = await this.prisma.user.create({
      data: {
        name: user.name,
        password: hashedPassword,
        roles: [Role.USER],
      },
    });

    return new UserResponse(userForDb);
  }

  async findOne(name: string) {
    return await this.prisma.user.findUnique({
      where: {
        name,
      },
    });
  }

  async hashPassword(password: string) {
    const hashedPassword = await bcrypt.hash(password, 10);
    return hashedPassword;
  }

  async generateAdminAccount() {
    const adminName = process.env.ADMIN_NAME;
    const adminFromDb = await this.prisma.user.findUnique({
      where: { name: adminName },
    });

    if (adminFromDb) {
      return true;
    }

    const hashedAdminPassword = await this.hashPassword(
      process.env.ADMIN_PASSWORD,
    );

    return !!(await this.prisma.user.create({
      data: {
        name: adminName,
        password: hashedAdminPassword,
        roles: [Role.ADMIN],
      },
    }));
  }
}
