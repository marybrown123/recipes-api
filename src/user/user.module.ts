import { Module } from '@nestjs/common';
import { UserService } from './user.service';
import { UserController } from './user.controller';
import { PrismaService } from '../prisma/prisma.service';
import { MailModule } from '../mail/mail.module';
import { TokenModule } from '../token/token.module';
import { WebhookModule } from '../webhook/webhook.module';

@Module({
  imports: [MailModule, TokenModule, WebhookModule],
  providers: [UserService, PrismaService],
  exports: [UserService],
  controllers: [UserController],
})
export class UserModule {}
