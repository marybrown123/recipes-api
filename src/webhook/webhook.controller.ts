import { Body, Controller, Get, Param, Patch, UseGuards } from '@nestjs/common';
import {
  ApiBadRequestResponse,
  ApiCreatedResponse,
  ApiForbiddenResponse,
  ApiOkResponse,
  ApiOperation,
  ApiParam,
  ApiTags,
  ApiUnauthorizedResponse,
} from '@nestjs/swagger';
import { UpdateWebhookDTO } from './DTOs/update-webhook.dto';
import { WebhookService } from './webhook.service';
import { WebhookResponse } from './responses/webhook.response';
import { AuthGuard } from '@nestjs/passport';
import { IsAdminGuard } from '../user/guards/is-admin.guard';

@ApiTags('webhook')
@Controller('/webhook')
export class WebhookController {
  constructor(private readonly webhookService: WebhookService) {}

  @Patch('/:id')
  @UseGuards(AuthGuard('jwt'), IsAdminGuard)
  @ApiOperation({ summary: 'Update a webhook' })
  @ApiCreatedResponse({ type: WebhookResponse })
  @ApiUnauthorizedResponse({ description: 'User not authorized' })
  @ApiForbiddenResponse({
    description: 'User is not an admin',
  })
  @ApiBadRequestResponse({ description: 'Bad request' })
  @ApiParam({ name: 'id', required: true })
  async updateWebhook(
    @Param('id') webhookId: number,
    @Body() newWebhook: UpdateWebhookDTO,
  ): Promise<WebhookResponse> {
    return this.webhookService.updateWebhook(webhookId, newWebhook);
  }

  @Get()
  @UseGuards(AuthGuard('jwt'), IsAdminGuard)
  @ApiOperation({ summary: 'Fetch all webhooks' })
  @ApiOkResponse({ type: [WebhookResponse] })
  @ApiUnauthorizedResponse({ description: 'User not authorized' })
  @ApiForbiddenResponse({
    description: 'User is not an admin',
  })
  async fetchAllWebhooks(): Promise<WebhookResponse[]> {
    return this.webhookService.fetchAllWebhooks();
  }
}
