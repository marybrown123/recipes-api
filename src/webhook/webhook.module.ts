import { Module } from '@nestjs/common';
import { WebhookService } from 'src/webhook/webhook.service';

@Module({
  providers: [WebhookService],
  exports: [WebhookService],
})
export class WebhookModule {}
