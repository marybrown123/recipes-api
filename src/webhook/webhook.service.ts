import { HttpService } from '@nestjs/axios';
import { Injectable, NotFoundException } from '@nestjs/common';
import { lastValueFrom } from 'rxjs';
import { UpdateWebhookDTO } from './DTOs/update-webhook.dto';
import { PrismaService } from '../prisma/prisma.service';
import { WebhookResponse } from './responses/webhook.response';
import { WebhookName } from './enums/webhookName.enum';

@Injectable()
export class WebhookService {
  constructor(
    private readonly httpService: HttpService,
    private readonly prismaService: PrismaService,
  ) {}
  async sendWebhook<T>(
    webhookPayload: T,
    webhookName: WebhookName,
  ): Promise<void> {
    const webhookFromDb = await this.fetchOneWebhookByName(webhookName);

    if (webhookFromDb.isEnabled) {
      try {
        await lastValueFrom(
          this.httpService.post(webhookFromDb.url, webhookPayload),
        );
      } catch (error) {
        throw new Error('There was an error while sending request');
      }
    }
  }

  async updateWebhook(
    webhookId: number,
    newWebhook: UpdateWebhookDTO,
  ): Promise<WebhookResponse> {
    const webhookToUpdate = await this.prismaService.webhook.findFirst({
      where: { id: webhookId },
    });

    if (!webhookToUpdate) {
      throw new NotFoundException();
    }

    const updatedWebhook = await this.prismaService.webhook.update({
      where: { id: webhookId },
      data: {
        ...newWebhook,
      },
    });

    return new WebhookResponse(updatedWebhook);
  }

  async fetchAllWebhooks(): Promise<WebhookResponse[]> {
    const webhooks = await this.prismaService.webhook.findMany();
    return webhooks.map((webhook) => {
      return new WebhookResponse(webhook);
    });
  }

  async fetchOneWebhookByName(
    webhookName: WebhookName,
  ): Promise<WebhookResponse> {
    const webhookFromDb = await this.prismaService.webhook.findFirst({
      where: { name: webhookName },
    });
    return new WebhookResponse(webhookFromDb);
  }
}
