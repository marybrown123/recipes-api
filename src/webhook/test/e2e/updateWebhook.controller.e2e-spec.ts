import { HttpStatus, INestApplication, ValidationPipe } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { Test, TestingModule } from '@nestjs/testing';
import { Role, User } from '@prisma/client';
import { AppModule } from '../../../app.module';
import { PrismaService } from '../../../prisma/prisma.service';
import { UserService } from '../../../user/user.service';
import { UpdateWebhookDTO } from '../../DTOs/update-webhook.dto';
import * as request from 'supertest';
import { WebhookService } from '../../webhook.service';
import { WebhookServiceMock } from '../mock/webhook.service.mock';

describe('Webhook Controller - Update', () => {
  let app: INestApplication;
  let testUser: User;
  let testAdmin: User;
  let userService: UserService;
  let prismaService: PrismaService;
  let jwtService: JwtService;
  let testAdminAuthenticationToken: string;
  let testUserAuthenticationToken: string;
  const correctPayload: UpdateWebhookDTO = {
    url: 'testUrl',
    isEnabled: false,
    retriesAmount: 5,
  };
  const validationTestStructure = [
    {
      property: 'url',
      value: 1,
    },
    {
      property: 'isEnabled',
      value: 1,
    },
    {
      property: 'retriesAmount',
      value: 'test',
    },
  ];

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    })
      .overrideProvider(WebhookService)
      .useClass(WebhookServiceMock)
      .compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(new ValidationPipe({ transform: true }));
    await app.init();
    userService = moduleFixture.get<UserService>(UserService);
    prismaService = moduleFixture.get<PrismaService>(PrismaService);
    jwtService = moduleFixture.get<JwtService>(JwtService);

    testUser = await userService.generateAccount(
      process.env.TEST_EMAIL,
      process.env.TEST_NAME,
      process.env.TEST_PASSWORD,
      Role.USER,
    );

    testAdmin = await userService.generateAccount(
      process.env.TEST_ADMIN_EMAIL,
      process.env.TEST_ADMIN_NAME,
      process.env.TEST_ADMIN_PASSWORD,
      Role.ADMIN,
    );

    await userService.updateVerificationStatus(testUser.id);
    await userService.updateVerificationStatus(testAdmin.id);

    testAdminAuthenticationToken = jwtService.sign({
      name: testAdmin.name,
      sub: testAdmin.id,
    });

    testUserAuthenticationToken = jwtService.sign({
      name: testUser.name,
      sub: testUser.id,
    });
  });

  afterAll(async () => {
    await prismaService.user.delete({ where: { id: testUser.id } });
    await app.close();
  });

  it('should update a webhook', async () => {
    return request(app.getHttpServer())
      .patch('/webhook/1')
      .send(correctPayload)
      .set('Authorization', `bearer ${testAdminAuthenticationToken}`)
      .expect(HttpStatus.OK);
  });

  it('should throw an error when user is not admin', async () => {
    return request(app.getHttpServer())
      .patch('/webhook/1')
      .send(correctPayload)
      .set('Authorization', `bearer ${testUserAuthenticationToken}`)
      .expect(HttpStatus.FORBIDDEN);
  });

  validationTestStructure.map((object) => {
    const uncorrectPayload = { ...correctPayload };
    uncorrectPayload[object.property] = object.value;
    return it(`should throw validation error when property ${object.property} is ${object.value}`, async () => {
      return request(app.getHttpServer())
        .patch('/webhook/1')
        .send(uncorrectPayload)
        .set('Authorization', `bearer ${testAdminAuthenticationToken}`)
        .expect(HttpStatus.BAD_REQUEST)
        .then((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });
  });
});
