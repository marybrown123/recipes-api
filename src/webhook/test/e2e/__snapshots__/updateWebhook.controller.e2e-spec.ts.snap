// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webhook Controller - Update should throw validation error when property isEnabled is 1 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "isEnabled must be a boolean value",
  ],
  "statusCode": 400,
}
`;

exports[`Webhook Controller - Update should throw validation error when property retriesAmount is test 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "retriesAmount must not be greater than 10",
    "retriesAmount must not be less than 1",
    "retriesAmount must be a number conforming to the specified constraints",
  ],
  "statusCode": 400,
}
`;

exports[`Webhook Controller - Update should throw validation error when property url is 1 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "url must be a URL address",
    "url must be a string",
  ],
  "statusCode": 400,
}
`;
